### DISCUSSION TITLE
How to ignore page rotation while signing pdf using pyHanko?

### DISCUSSION THREAD

**Discussion by ihegde** [MatthiasValvekens/pyHanko#150] [Q&A] [answered]
hi, I'm trying to use pyHanko to add a visible signature to a PDF document. One of the documents has its page rotated by 90. 

I'm using the following code from the samples. I don't need background, just plain text. 
The problem is, the page is rotated ( but it's visually alright, they've just done it to get more horizontal space) But the signature is appearing vertical! Is there any way I can ask pyHanko to sign straight ignoring the page rotation? 

    fields.append_signature_field(
        w, sig_field_spec=fields.SigFieldSpec(
            'Signature', box=(330, 229, 530, 289)  # width 200 , height 60
        )
    )

    meta = signers.PdfSignatureMetadata(field_name='Signature')
    pdf_signer = signers.PdfSigner(
        meta, signer=signer, stamp_style=stamp.TextStampStyle(
            # the 'signer' and 'ts' parameters will be interpolated by pyHanko, if present
            stamp_text='Digitally signed by\n%(signer)s\nTime: %(ts)s',
            text_box_style=text.TextBoxStyle(
                font=opentype.GlyphAccumulatorFactory('NotoSans-Regular.ttf')
            ),
            #background=images.PdfImage('stamp.png')
        ),
    )

**Comment by MatthiasValvekens** [comment, marked as answer]
Hi @ihegde,

Per the PDF standard, the default behaviour of the **Rotate** entry in a page dictionary is to rotate all annotations along with the page content. That also applies to (visible) signature fields, and I'm pretty sure that's the cause of the behaviour you're seeing.

The behaviour you want can be achieved by setting the `NoRotate` flag on the annotation, but unfortunately there's no convenient API to set that bit for you in pyHanko. It's not hard to add an option to `SigFieldSpec` to toggle that one, though. I'll see what I can do about that and get back to you.

In the meantime, you can work around the issue as follows:

 - Set up a signature field using `pyhanko.sign.fields.append_signature_field(...)`, in much the same way as you do now.
 - Change the field's annotation flags manually (see snippet below for an example)
 - Sign.

Flag snippet:
```
flags = w.root['/AcroForm']['/Fields'][0]['/F']   # assuming there's only one field
w.root['/AcroForm']['/Fields'][0]['/F'] = generic.NumberObject(flags | 0b10000)
```

(warning: untested code)


Hope that helps.

**Reply by ihegde** [reply]
It worked! thanks. That would be great if you can provide NoRotate flag. Thanks again for this library and the workaround.

**Reply by MatthiasValvekens** [reply]
Dealt with in d99a91f. I was planning to do a release soon anyhow; this will make it in.

**Reply by ihegde** [reply]
@MatthiasValvekens Awesome! Thanks so much.

**Reply by ihegde** [reply]
@MatthiasValvekens Observed a strange behavior on Adobe reader DC. 
I tried testing the visibility and position on different readers/editors. 
With PDF-XChange Editor, the signature appearance comes up perfectly. 
With Adobe Reader DC,, it still shows vertically. 
Just to explore a bit further, I compared the flags: 
when signed with pyHanko:
\F : 148 = 1001 0100  (noRotate flag is set, Rect : 494 510 699 580)

If I manually sign the same PDF with adobe reader, (by drawing the position with mouse) 
\F :132 = 1000 0100 (noRotate flag is not set, Rect : 509 586 551 761 )
I can understand this because as you said this is the default behavior to rotate with the content. 

Printing is working fine from both readers! 

Why would Adobe reader and PDF-XChange draw the same content differently when they are all supposed to follow the same standard? Is there any other flag I need to set?

**Reply by MatthiasValvekens** [reply]
Hmm, that's something I'd have to investigate. My schedule is quite packed the coming days, but I'll see what I can find out.

As a last resort, if there's indeed a significant difference in implementation (which there shouldn't be), you could always remove the `Rotate` entry and manually rotate all page graphics by prepending a `<rotation matrix goes here> cm` instruction to the content stream. It's a hack, and it won't work if there are prior signatures on your document, but maybe it works for your use case.

**Reply by ihegde** [reply]
@MatthiasValvekens , Thanks for the quick response and the hack. Will explore this but some pdfs need multiple signatures on the same document, so I'll continue to explore/ask if the the user can use PDF-XChange.

**Reply by ihegde** [reply]
@MatthiasValvekens, 
Slight progress: 
I compared each field of the annotation when signed manually by PDF XChange editor and pyHanko: 
Manual signing: 
{'AP': {'N': <PDFObjRef:9>}, 'F': 4, 'FT': /'Sig', 'MK': {'R': 90}, 'P': <PDFObjRef:8>, 'Rect': [511.240875, 595.36554, 543.240845, 745.36554], 'Subtype': /'Widget', 'T': b'Signature1', 'Type': /'Annot', 'V': <PDFObjRef:10>}

pyHanko: (Default settings) 
{'FT': /'Sig', 'T': b'Signature1', 'Type': /'Annot', 'Subtype': /'Widget', 'F': 132, 'Rect': [509, 586, 551, 761], 'P': <PDFObjRef:6>, 'AP': {'N': <PDFObjRef:28>}, 'V': <PDFObjRef:29>}

There is only one difference: The manually signed annotation has this property:
 'MK': {'R': 90},
From the documentation, this determines the properties of the contents, and R:90 means the content has to rotate by 90!  
So I added this to my code as below: 

        w.root['/AcroForm']['/Fields'][0]['/MK'] = generic.DictionaryObject({pdf_name('/R'): generic.NumberObject(90)})

Now, the pyHanko signed PDF annotation has the following structure: 
{'FT': /'Sig', 'T': b'Signature1', 'Type': /'Annot', 'Subtype': /'Widget', 'F': 132, 'Rect': [509, 586, 551, 761], 'P': <PDFObjRef:6>, 'AP': {'N': <PDFObjRef:28>}, 'MK': {'R': 90}, 'V': <PDFObjRef:29>}

I thought this is it, but the readers are still showing the signature vertically :(  ( Box comes up in the right place but the signature text inside the box is vertical) 
The only thing left to analyse is actual "/Contents" . Since it is stream, I couldn't interpret the differences if any. which field does "stamp_text" of TextStampStyle get into? Is there any formatting done with this?

**Reply by MatthiasValvekens** [reply]
The **MK** dictionary is only used to store hints for appearance _generation_ purposes. In other words, they're only useful when the viewer has to (re)generate the widget's appearance from scratch. Since (visible) signature widgets have a pregenerated appearance that is essentially unchanged, those values don't have all that much meaning for signatures. Either way, they shouldn't affect rendering at all.

I'll also do some experiments over the coming days, when I find the time.

**Reply by MatthiasValvekens** [reply]
> The only thing left to analyse is actual "/Contents" . Since it is stream, I couldn't interpret the differences if any. which field does "stamp_text" of TextStampStyle get into? Is there any formatting done with this?

Just to make sure we're looking at the same thing: the value under `['/V']['/Contents']` is the signature container, and has no bearing on the appearance. All the visual stuff is contained in the stream attached to `['/AP']['/N']`. That's a PDF graphics stream, but depending on how you're generating the appearances (both with pyHanko and with your favourite viewer(s)), that stream could be anywhere from very easy to very difficult to inspect "by hand".

If you want to do the investigation yourself, here's a very useful tool: https://github.com/itext/i7j-rups.

**Reply by MatthiasValvekens** [reply]
So, comparing how Acrobat generates signature appearances in rotated vs. unrotated documents, it's quite clear that it doesn't bother with `NoRotate` at all, and simply compensates for the rotation in the appearance stream.

Rotated doc signed w/ Acrobat:


```
q
0 1 -1 0 217.4188 0 cm
/n0 Do
Q
q
0 1 -1 0 217.4188 0 cm
/n2 Do
Q
```

Unrotated doc signed w/ Acrobat


```
q
1 0 0 1 0 0 cm
/n0 Do
Q
q
1 0 0 1 0 0 cm
/n2 Do
Q
```

As you can see, there's a nontrivial transformation matrix (the argument to `cm`) in the rotated file that compensates for the page rotation. In a regular file, that's just the identity transformation (i.e. a no-op).

It's not fully clear to me yet why `NoRotate` isn't behaving the way we expect it to (I've tried some non-Adobe viewers as well with similar results), but perhaps compensating for the rotation in the appearance stream is just easier. I'll sleep on it.

**Reply by ihegde** [reply]
Oh! does it mean it's not a hack after all and that's the only way? Adding multiple signatures using adobe reader works ok,
though we need to save it multiple times: 
1. Sign the document using first signature, save as signed1.pdf 
2. Open the Signed1.pdf, sign with second signature and save as Signed2.pdf 
I could do the same steps programmatically, hope it will not make the first signature invalid. That's how I'm adding signatures to multiple placeholders in the same document currently.

**Reply by MatthiasValvekens** [reply]
> Oh! does it mean it's not a hack after all and that's the only way?

The "hack" I mentioned further up in the thread was about embedding the rotation in the page content stream (thus removing the need to worry about `/Rotate` going forward). That will always invalidate existing signatures. This workaround is different: it rotates the _signature_ appearance only. That's fine and shouldn't cause any issues with prior signatures.

> Open the Signed1.pdf, sign with second signature and save as Signed2.pdf. I could do the same steps programmatically, hope it will not make the first signature invalid. That's how I'm adding signatures to multiple placeholders in the same document currently.

Yeah, that's expected and independent of the whole visual appearance/rotation discussion. PDF signatures are always serial like that. [Here's some background from the docs](https://pyhanko.readthedocs.io/en/latest/cli-guide/signing.html#pdf-signing-background) (well, there are many explanations out there on the internet, you don't have to take my word for it :) )

**Reply by ihegde** [reply]
Please pardon my ignorance on PDF intricacies. When I saw cm with matrix I thought both are the same. I'll study some more and understand. I was totally new to PDF when I started a couple of weeks ago. Thanks for your guidance and pointers without which I wouldn't have come this far :)

**Reply by MatthiasValvekens** [reply]
No problem, I should've been more clear in my explanation :).

**Reply by ihegde** [reply]
Hi @MatthiasValvekens, 
Update on this: I was finally able to get the appearance of signature working. Since I was not able to find a generic solution, I have hardcoded cm and Td in a couple of places.  I have limited variations of font sizes, box sizes, and also rotation (0 and 90) so works for my requirements. 
Thanks!
